105 submission detail

ashberg opened the pull request:
Fixes #105

ashberg committed:
Working on setting up the foundation for the submission_detail component.

ashberg committed:
Created the AGTestCommandResult component.

ashberg committed:
Created a style sheet for the submission_detail component.

ashberg committed:
Added the submission_detail_panel component.

ashberg committed:
Adding tests to test_ag_command_result.ts.

ashberg committed:
Working on tests in test_mutation_suite_result.ts.

ashberg committed:
Added tests to test_submission_detail.ts.

ashberg committed:
Cleaning up tests in the test_submission_detail directory.

ashberg committed:
Fixed submission_detail tests after adding Watchers. Fixed submission_detail_panel layout.

ashberg committed:
Adjusted the styling for the submission_detail component.

ashberg committed:
Updated the refresh_submissions method in the submission_list component to update the d_selected_submission when the status of that submission changes.

ashberg committed:
Adjusted the selected_submission_with_results Watcher test in test_submission_detailer.ts.

ashberg committed:
Added calls to the get_output_size methods in the ag_case_setup_result, ag_command_result and mutation_suite_result components. Stubbed the get_output_size methods for those components in testing and ensured they were called.

ashberg committed:
Adjusted tests concerning Watched properties in the submission_detail directory.

james-perretta committed:
Moved wrapped operator to next line.

james-perretta committed:
Updated order of items in adjust fdbk menu and added user-friendly text.

james-perretta committed:
Added collapsible sidebar to submission list.

ashberg committed:
Added grading status symbols to the grading status section. Changed the conditional for displaying the submission_score to be limited to when the status of the submission is (waiting_for_deferred or finished_grading) AND the total_points_possible for the submission is greater than 0. Removed the the call to get_submission_result from created() in the submission_detail component.

ashberg committed:
Updated the styling for the submission_detail_panel. Adjusted the background-color and border-color of the lengthy-output feedback sections. Added a border around the diff components in ag_command_result. Added icons to express correctness/timed-out in the ag_command_result and mutation_suite_result components. Changed the order in which feedback appears in the ag_command_result component.

ashberg committed:
Moved the ReturnCodeCorrectness enum into its own file. Added timed-out/not-available icons to the ag_case_setup_result component.

ashberg committed:
Removed unused css classes from the ag_command_result component.

ashberg committed:
Experimenting with a new layout design for the mutation_suite_result, ag_case_setup_result, and ag_command_result components.

ashberg committed:
Added { position: sticky; top: 0 } to the panel-header-open class in the submission_detail_component. Removed the feedback-labels from the diff section in the ag_command_result component.

ashberg committed:
Changed the way submission-detail-panel output is displayed on smaller screens. Changed the lengthy-output max-height and diff height to 50vh. Changed the watcher in submission_detail to set d_submission_result to the new_value's results. Changed all labels from 'stdout' and 'stderr' to 'Output' and 'Error Output' respectively.

ashberg committed:
Removed an element from the submission_detail html template that I was using for testing. Removed the commented out code from the submission_detail_panel css section.

ashberg committed:
Removed the z-index setting from the command-panel-header-open class. Decremented the z-index assigned to the panel-header-open class. Added the z-index property to the submission_list's sidebar-menu class and set it equal to one more than the panel-header-open z-index.

ashberg committed:
The submission_detail component now uses a getter for accessing the results of the submission being viewed.

ashberg committed:
Renamed d_submission_result to d_submission_fdbk_override in the submission_detail component.

ashberg committed:
Added conditions in both the ag_case_result and ag_suite_result components to perceive the output correctness of a command as correct if the stdout or stdout are to be shown, both stdout and stderr aren't worth any points and stdout and stderr correct are both null.

ashberg committed:
Adjusted the test cases in test_ag_case_result.ts to accomodate the output_only CorrectnessLevel that can now be attributed to a test case panel in the ag_case_result component.

ashberg committed:
Adjusted the test cases in test_ag_suite_result.ts to accomodate the info_only correctness level that could be attributed to the case panel. Changed the CorrectnessLevel.output_only to be CorrectnessLevel.info_only.

ashberg committed:
Added an icon for the not_available CorrectnessLevel in the CorrectnessLevel component.

ashberg committed:
Fixed the value assigned to the name property for the submission-detail-panel in the ag_case_result component.

ashberg committed:
Added the border around the mutation and ag-test-suites back. Adjusted the color of the info-only submission-detail-panel. Renamed the variables in the style portion of the submission_detail_panel component to reflect the correctness level instead of the color currently assigned to that correctness level panel. Removed the not-available icon from the correctness-level component.

ashberg committed:
Removed the call to get_results() from the fdbk_category and submission Watchers in the ag_command_result, ag_case_setup_result and mutation_suite_result components. Changed the location of the d_loaded variables to before any calls to load the result content for the panel so the spinning icon doesn't show up at different times while fetching content. Added buttons in the mutation_suite_result component to hide/show section output and tested that implementation accordingly.

ashberg committed:
Changed the layout for the show/hide output sections in the mutation_suite_result component.

ashberg committed:
Removed console.log statements from the submission_detail component.

ashberg committed:
Changed the 'feedback-explanation' text for the 'invalid-tests-section' in the mutation_suite_result component.

ashberg committed:
Removed the assignment of hardcoded values into invalid_tests in the mutation_suite_result component.

ashberg committed:
Fixed a trailing whitespace error in the mutation_suite_result component.

ashberg committed:
Rearranged the feedback sections in the mutation_suite_result component. Updated test cases to reflect the new arrangement.

ashberg committed:
Changed the order in which the student tests lists are displayed in the mutation_suite_result component.

ashberg committed:
Updated the icons used in the 'student tests' section of the mutation_suite_result component.

ashberg committed:
Changed the color of the none_correct correctness icon to be lighter. Changed the v-if condition for the exit status in the mutation_suite_result, ag_command_result, and ag_case_setup_result components to only show the exit status if the return_code is not null/not_available. Replaced the correctness icons in the submission_detail results components with the same icons but with the circle removed. Removed the 'actual and expected return code' tooltip from the ag_command_result component and displayed the feedback previously in the tooltip as feedback-rows.

ashberg committed:
Changed the html that displays the lists of tests and bugs so that when the page narrows the list item stays on the same line as the list icon.

ashberg committed:
Working on testing the get_mutation_test_validity_correctness_level, suite_bugs_exposed_correctness, and suite_student_test_correctness, and suite_return_code_correctness functions from the mutation_suite_results component.

ashberg committed:
Added tests for the mutation_suite_correctness function from the MutationSuiteResults component. Adjusted MutationSuiteResult tests to accomodate the student tests fieldset always being visible.

ashberg committed:
Added classes to apply colors indicative of the actual_return_code's correctness to the actual_return_code feedback.

ashberg committed:
Adjusted mutation_suite_results tests to use the set return_code/points/student_tests correctness functions. Adjusted mutation_suite_result tests to check for the existence of the no-tests-submitted-message.

ashberg committed:
Removed comments from the setup_return_code_correctness, student_tests_correctness, and points_for_bugs_exposed_correctness functions in mutation_suite_results.vue.

ashberg committed:
Adjusted comments in the mutation_suite_correctness function in the MutationSuiteResults component.

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+          Since you ran out of late days, this submission will not count towards your final grade.
+        </p>
+        <div v-if="!is_group_member">
+          <div v-if="d_submission.does_not_count_for.length">
```
Let's combine these two with &&

ashberg committed:
Added a setup_return_code_correctness helper function in correctness.ts.

ashberg committed:
Removed the CorrectnessLevel import from the top of mutation_suite_results.vue.

ashberg committed:
Adjusted the points_for_bugs_exposed_correctness function in mutation_suite_results and any tests in test_mutation_suite_results affected by the changes.

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+        <div v-if="(d_submission.status === GradingStatus.waiting_for_deferred
+                   || d_submission.status === GradingStatus.finished_grading) &&
+                   submission_result !== null
+                   && submission_result.total_points_possible !== 0"
```
Add 1 space to beginning of line 87.
`&&` should go on next line.

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+                 class="fas fa-list being-graded-symbol"></i>
+              <i v-else-if="d_submission.status === GradingStatus.waiting_for_deferred
+                            || d_submission.status === GradingStatus.finished_grading"
+                 class="far fa-check-circle finished-grading-symbol"></i>
```
Let's hide `#grading-status-section` and `#grading-status` when the status is waiting_for_deferred or finished_grading and the score is available.

For staff, let's add a message to the score that says "Deferred tests pending" for `waiting_for_deferred` and "Deferred tests finished" for `finished_grading`

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+          on <span id="submission-timestamp">{{format_datetime(d_submission.timestamp)}}</span>
+        </div>
+
+        <div v-if="being_processed"
```
Let's also hide this message non-staff and `waiting_for_deferred` submissions. (Don't modify being_processed, just add to this if):
`v-if="being_processed || (!d_globals.user_roles.is_staff && d_submission.status === GradingStatus.waiting_for_deferred)`

ashberg replied:
I don't think adding this || (!d_globals.user_roles.is_staff && d_submission.status === GradingStatus.waiting_for_deferred) changes anything? being_processed will already return true if status === waiting_for_deferred.

ashberg replied:
Should I change being_processed to return false when status === 'waiting_for_deferred'?

james-perretta replied:
Hmmm good point. Making that change to being_processed doesn't seem quite right unless we named it something else...

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+                   submission_result !== null
+                   && submission_result.total_points_possible !== 0"
+             id="submission-score"> Score:
+          {{submission_result.total_points}}/{{submission_result.total_points_possible}}
```
Let's make the points bold

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+              Don't worry, it's probably our fault, not yours!
+              Sometimes these things only happen once, so you might as well submit again.
+              We won't count this one towards your daily limit.
+              If this keeps happening, send us an email and we'll look into it!
```
New text (make the email address bold, not code-formatted):
An unexpected error occurred while grading your submission. This submission will not count towards your daily limit. Please wait a few minutes and try your submission again. If the problem persists, please contact `help@autograder.io` and include the information **"Submission ID: {d_submission.pk}"** in your email.

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+          </div>
+        </div>
+
+        <div v-if="(d_submission.status === GradingStatus.waiting_for_deferred
```
Let's put score right below the "Submitted by" header rather than below all the info/status messages

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+        <p v-if="is_group_member && does_not_count_for_current_user"
+           id="does-not-count-for-user-message">
+          Since you ran out of late days, this submission will not count towards your final grade.
+        </p>
```
Can we figure out a way to differentiate some of these info messages from each other?
This applies to the bonus submission and late day/does not count for text.

One possibility would be to pick a different color for the info messages, make the spacing between them a bit more consistent (e.g. a consistent margin top and bottom rather than mixing div and p) and add some kind of "info" icon at the beginning of the message.

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+
+          <div id="grading-status">
+            <div v-if="d_submission.status === GradingStatus.received">
+              We got your submission! It should be queued soon.
```
Can we make the status text a different color, perhaps stormy-gray-dark?

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+        </div>
+      </div>
+
+      <div v-if="submission_result !== null">
```
Can submission_result ever be null?

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+  d_globals = this.globals;
+
+  @Prop({required: true, type: Object})
+  selected_submission_with_results!: SubmissionWithResults;
```
Rename to `submission_with_results`

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+  @Watch('selected_submission_with_results')
+  on_selected_submission_change(new_value: SubmissionWithResults,
+                                old_value: SubmissionWithResults) {
+    this.d_loading = true;
```
Nothing async happens here, so we don't need to toggle loading

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+                                old_value: SubmissionWithResults) {
+    this.d_loading = true;
+    this.d_submission = new Submission(new_value);
+    this.d_fdbk_category = this.determine_feedback_type();
```
Instead of having to recompute these variables manually, let's replace them with computed properties like so:

```
get submission() {
    return new Submission(this.submission_with_results);
}
```

~~For d_fdbk_category, just make `determine_feedback_type` a getter and rename it to fdbk_category.~~

Then, we can replace uses of d_submission with submission. ~~and uses of d_fdbk_category with fdbk_category.~~

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+  d_submission: null | Submission = null;
+  d_submission_fdbk_override: null | SubmissionResultFeedback = null;
+  d_loading = true;
+  d_height_of_view_file = "500px";
```
Does this value ever change? Let's just inline the value on line 114 (the multi-file-viewer inputs). 
Let's also make this something like 50vh or 75vh, similar to what we did with the output height.

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+  created() {
+    this.d_submission = new Submission(this.selected_submission_with_results);
+    this.d_user = this.d_globals.current_user;
+    this.d_user_roles = this.d_globals.user_roles;
```
As a general rule, don't alias values in d_globals. By doing so, we lose reactivity for those members. We also want it to be clear that we're accessing global data.

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+    this.d_user_roles = this.d_globals.user_roles;
+    this.d_fdbk_category = this.determine_feedback_type();
+    this.d_submission_fdbk_override = this.selected_submission_with_results.results;
+    this.d_loading = false;
```
Nothing async happens here, so we don't need to mark as loading/finished loading.
Come to think of it, it looks like we don't need to wrap the whole component in loading at all.

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+            && this.d_submission!.status !== GradingStatus.removed_from_queue);
+  }
+
+  async load_results() {
```
In this function, we should toggle some sort of loading variable. 
Then, while this is loading, we should disable the adjust feedback selector and a loading spinner next to it or `cursor: wait` when hovering over it.

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+
+  async open_file(filename: string) {
+    let multi_file_viewer = <MultiFileViewer> this.$refs.view_submission_result_files;
+    let file_content: Promise<string> = Promise.resolve(
```
Doesn't get_file_content already return a promise?

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+    FileSaver.saveAs(new File([await this.d_submission!.get_file_content(filename)], filename));
+  }
+
+  async remove_submission_from_queue() {
```
Let's add a `d_removing_from_queue` bool, toggle it here, and disable the remove from queue button (the one in the modal) while true.

james-perretta replied:
We also need some API error handling here in case the submission just started to be grading or someone else removed it from the queue (this was missing from the old versions too I think).

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+
+  async remove_submission_from_queue() {
+    await this.d_submission!.remove_from_queue();
+    this.d_submission!.status = GradingStatus.removed_from_queue;
```
`.remove_from_queue()` should automatically update `.status`

james-perretta commented on src/components/project_view/submission_detail/submission_detail_panel.vue:
```diff
@@ -0,0 +1,290 @@
+<template>
```
What if we named this component something like "result panel" instead?

james-perretta commented on src/components/project_view/submission_detail/submission_detail_panel.vue:
```diff
@@ -0,0 +1,290 @@
+            </correctness-icon>
+          </div>
+      </template>
+
```
Remove this blank line

james-perretta commented on src/components/project_view/submission_detail/submission_detail_panel.vue:
```diff
@@ -0,0 +1,290 @@
+    all_correct = 'all-correct',
+    info_only = 'info-only',
+    output_only = 'output-only',
+    dont_check = 'dont-check'
```
Do these two ever get used?

james-perretta commented on src/components/project_view/submission_detail/submission_detail_panel.vue:
```diff
@@ -0,0 +1,290 @@
+
+import CorrectnessIcon, { CorrectnessLevel } from "@/components/project_view/submission_detail/correctness_icon.vue";
+
+export enum PanelCorrectnessLevel {
```
I think we have 2 definitions of pretty much the same enum. Can we move it into correctness.ts?
I'd say keep this version with the hyphens and just add a comment that the strings are hyphenated for easier interpolation in CSS class names.

james-perretta commented on src/components/project_view/submission_detail/submission_detail_panel.vue:
```diff
@@ -0,0 +1,290 @@
+  is_multi_command_case!: boolean;
+
+  d_correctness_level: CorrectnessLevel = CorrectnessLevel.none_correct;
+  d_panel_is_active = false;
```
Since we never need to modify this internally, we can get rid of it and just use `panel_is_active` instead

james-perretta commented on src/components/project_view/submission_detail/submission_detail_panel.vue:
```diff
@@ -0,0 +1,290 @@
+  d_is_open = false;
+
+  @Watch('panel_is_active')
+  on_panel_is_active_change(new_value: boolean, old_value: boolean) {
```
Does this value only ever potentially change when the user changes the adjust feedback setting? If that's the case, I'd argue that we don't want to watch panel_is_active so that whatever panels the user has open/closed stay that way when they change the feedback.

If all of that is the case and we make that change, then I think we should rename this variable to something like `open_initially`

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,502 @@
+
+        <div v-for="(ag_test_suite_result, index) of submission_result.ag_test_suite_results"
+             ref="ag_test_suite_results">
+          <AGTestSuiteResult :submission="d_submission"
```
Add `:key="ag_test_suite_result.pk`

james-perretta commented on src/components/project_view/submission_detail/submission_detail_panel.vue:
```diff
@@ -0,0 +1,290 @@
+  @Prop({default: false, type: Boolean})
+  is_multi_command_case!: boolean;
+
+  d_correctness_level: CorrectnessLevel = CorrectnessLevel.none_correct;
```
Since we never need to modify this internally, we can get rid of it and just use `correctness_level` instead

james-perretta commented on src/components/project_view/submission_detail/submission_detail_panel.vue:
```diff
@@ -0,0 +1,290 @@
+    return PanelCorrectnessLevel.some_correct;
+  }
+
+  toggle_d_is_open() {
```
Here's some code to add to fix some shenanigans from opening/closing with sticky panels:

At the beginning, add:
`let top = this.$el.getBoundingClientRect().top;`

At the end, add:
```
    // This prevents any open panels below this one from being pushed
    // into the top of the viewport due to the size change of the parent.
    if (!this.d_is_open && top < 0) {
      this.$nextTick(() => {
        this.$el.scrollIntoView();
      });
    }
```

james-perretta commented on src/components/project_view/submission_detail/submission_detail_panel.vue:
```diff
@@ -0,0 +1,290 @@
+  border: 1px solid $border-color;
+  font-weight: bold;
+  cursor: $cursor-option;
+}
```
Here's a fun scss trick you can do (you don't have to make this change, just letting you know it's a possibility):

```
@mixin panel_header($background-color: $white-gray, $border-color: $white-gray,
                    $cursor-option: pointer) {
  ...

  // The & gives you the parent selector
  &:hover {
    background-color: darken($background-color, 5%);
    border-color: darken($border-color, 5%);
  }
}
```

Then you don't need the second mixin or the explicit :hover classes later on

ashberg committed:
Made the changes requested on the pull request for branch 105 pertaining to the submission_detail component.

james-perretta commented on src/components/project_view/submission_detail/mutation_suite_results.vue:
```diff
@@ -0,0 +1,170 @@
+      return CorrectnessLevel.none_correct;
+    }
+
+    if (!suite_result.fdbk_settings.show_points && suite_result.num_bugs_exposed > 0) {
```
Let's try to avoid looking at the fdbk settings directly, as we can usually get more info at once from looking at aspects of the result. For example, here we'd want to check if total_points_possible is 0 rather than whether points are shown because there are other things that can cause total_points_possible to be 0 besides points being hidden.

A notable exception to this guideline is determining whether output is available in some cases.

ashberg committed:
Changed the color of the grading-status text in the SubmissionDetail component.

james-perretta commented on src/components/project_view/submission_detail/mutation_suite_results.vue:
```diff
@@ -0,0 +1,170 @@
+    return CorrectnessLevel.some_correct;
+  }
+
+  mutation_suite_correctness(suite_result: MutationTestSuiteResultFeedback) {
```
Suggested implementations for this method, `student_tests_correctness`, and `points_for_bugs_exposed_correctness`:

```
  student_tests_correctness(suite_result: MutationTestSuiteResultFeedback) {
    if (suite_result.invalid_tests === null) {
      return suite_result.discarded_tests.length === 0
             ? CorrectnessLevel.not_available : CorrectnessLevel.some_correct;
    }

    if (suite_result.invalid_tests.length === 0 && suite_result.discarded_tests.length === 0) {
      return CorrectnessLevel.all_correct;
    }

    if (suite_result.invalid_tests.length === suite_result.student_tests.length) {
      return CorrectnessLevel.none_correct;
    }

    return CorrectnessLevel.some_correct;
  }

  points_for_bugs_exposed_correctness(suite_result: MutationTestSuiteResultFeedback) {
    if (suite_result.num_bugs_exposed === null) {
      return CorrectnessLevel.not_available;
    }

    if (suite_result.num_bugs_exposed === 0) {
      return CorrectnessLevel.none_correct;
    }

    let total_points = typeof(suite_result.total_points) === 'string'
        ? parseFloat(suite_result.total_points) : suite_result.total_points;
    let total_points_possible = typeof(suite_result.total_points_possible) === 'string'
        ? parseFloat(suite_result.total_points_possible) : suite_result.total_points_possible;

    if (total_points_possible === 0 && suite_result.num_bugs_exposed > 0) {
      return CorrectnessLevel.info_only;
    }

    if (total_points === total_points_possible) {
      return CorrectnessLevel.all_correct;
    }
    return CorrectnessLevel.some_correct;
  }

  mutation_suite_correctness(suite_result: MutationTestSuiteResultFeedback) {
    let return_code_correctness = this.setup_return_code_correctness(suite_result);
    let points_correctness = this.points_for_bugs_exposed_correctness(suite_result);
    let student_tests_correctness = this.student_tests_correctness(suite_result);

    // received no points
    if (points_correctness === CorrectnessLevel.none_correct) {
      return CorrectnessLevel.none_correct;
    }

    let correctnesses = [return_code_correctness, points_correctness, student_tests_correctness];
    correctnesses = correctnesses.filter(val => val !== CorrectnessLevel.not_available);

    if (correctnesses.length === 0) {  // All 3 were not_available
      return CorrectnessLevel.not_available;
    }

    correctnesses = correctnesses.filter(val => val !== CorrectnessLevel.info_only);
    if (correctnesses.length === 0) {  // All remaining were info_only
      return CorrectnessLevel.info_only;
    }

    if (correctnesses.every(val => val === CorrectnessLevel.all_correct)) {
        return CorrectnessLevel.all_correct;
    }

    if (correctnesses.every(val => val === CorrectnessLevel.none_correct)) {
        return CorrectnessLevel.none_correct;
    }

    // We have a mix of all, some, and none correct
    return CorrectnessLevel.some_correct;
  }
```

james-perretta replied:
With regard to the tests, I don't think having one test for each explicit case is feasible anymore now that bugs exposed correctness can be info_only. A more sustainable approach might be to start by maximizing branch coverage and then add whatever meaningful edge cases that doesn't cover.

ashberg committed:
Adjusted styling of the grading-text in SubmissionDetail. Added a test to make sure that the api_error component exists when the remove_from_queue call is not successful.

james-perretta commented on src/components/project_view/submission_detail/mutation_suite_result.vue:
```diff
@@ -0,0 +1,797 @@
+           || this.d_mutation_test_suite_result!.fdbk_settings.show_setup_stderr
+           || this.d_mutation_test_suite_result!.setup_return_code !== null
+           || (this.d_mutation_test_suite_result!.setup_timed_out !== null
+               && this.d_mutation_test_suite_result!.setup_timed_out!);
```
Wow, the type-checker is really persnickety about this. For a shorter version that will appease it, try: `!!this.d_mutation_test_suite_result!.setup_timed_out`.

Yay javascript....

ashberg replied:
When I try this approach I get the warning: "This type is not allowed in the operand for the '!' operator because it could be null. Only booleans are allowed."

james-perretta replied:
Huh, It was fine when I tried it. I might have done something slightly differently. Go ahead and leave it as is then

james-perretta commented on src/components/project_view/submission_detail/mutation_suite_result.vue:
```diff
@@ -0,0 +1,797 @@
+
+        <div class="feedback-row"
+             v-if="d_mutation_test_suite_result.setup_return_code !== null
+                   || (d_mutation_test_suite_result.setup_timed_out !== null
```
See comment in `show_setup_fieldset`

james-perretta commented on src/components/project_view/submission_detail/mutation_suite_result.vue:
```diff
@@ -0,0 +1,797 @@
+
+    <fieldset v-if="show_setup_fieldset"
+              class="fieldset">
+      <legend class="legend"> Setup Command </legend>
```
Could we use the setup_command_name here instead (and fall back to just "Setup")

james-perretta commented on src/components/project_view/submission_detail/mutation_suite_result.vue:
```diff
@@ -0,0 +1,797 @@
+          <div class="feedback">
+            <div class="short-output">{{d_mutation_test_suite_result.setup_return_code}}</div>
+          </div>
+        </div>
```
Assuming we change the legend text, could we combine these two rows somehow?

james-perretta replied:
Maybe we should just take the same approach as in ag_case_setup_result.vue and just show the raw exit status. I'd also be ok with putting the exit status and then an icon right next to it.

james-perretta commented on src/components/project_view/submission_detail/mutation_suite_result.vue:
```diff
@@ -0,0 +1,797 @@
+            <template v-else>
+              <div v-if="!d_setup_stdout_content" class="short-output"> No output </div>
+              <pre v-else class="lengthy-output">{{d_setup_stdout_content}}</pre>
+            </template>
```
I don't suppose it would be worth writing a small component to handle output loading and displaying?
Something like what ViewFile does?

james-perretta commented on src/components/project_view/submission_detail/mutation_suite_result.vue:
```diff
@@ -0,0 +1,797 @@
+             class="feedback-row">
+
+          <div class="feedback-explanation">
+            Tests with false positives incorrectly report a bug when run against a correct
```
Word tweak: report -> reported

james-perretta commented on src/components/project_view/submission_detail/mutation_suite_result.vue:
```diff
@@ -0,0 +1,797 @@
+  d_validity_check_stderr_loaded = false;
+  d_grade_buggy_stdout_loaded = false;
+  d_grade_buggy_stderr_loaded = false;
+  d_mutation_test_suite_result_output_size: ResultOutput.MutationTestSuiteResultOutputSize
```
It would be ok to use a shorter name like `d_output_size` here

ashberg committed:
Made the changes requested on the pull request for branch 105 pertaining to the result_panel component.

ashberg committed:
Removed the panel_header_hover mixin.

james-perretta commented on src/components/project_view/submission_detail/ag_suite_result.vue:
```diff
@@ -0,0 +1,283 @@
+    </submission-detail-panel>
+
+    <template v-for="ag_test_case_result of d_ag_test_suite_result.ag_test_case_results">
+      <submission-detail-panel
```
Add `:key="ag_test_case_result.pk"`

james-perretta commented on src/components/project_view/submission_detail/ag_suite_result.vue:
```diff
@@ -0,0 +1,283 @@
+  d_fdbk_category: FeedbackCategory = FeedbackCategory.past_limit_submission;
+
+  @Watch('ag_test_suite_result')
+  on_ag_test_suite_result_change(new_value: object, old_value: object) {
```
Let's use a more specific type here instead of `object`.

james-perretta commented on src/components/project_view/submission_detail/ag_suite_result.vue:
```diff
@@ -0,0 +1,283 @@
+
+  @Watch('ag_test_suite_result')
+  on_ag_test_suite_result_change(new_value: object, old_value: object) {
+    this.d_ag_test_suite_result = JSON.parse(JSON.stringify(new_value));
```
Why do we need a deep copy of this? If this component doesn't mutate `d_ag_test_suite_result`, then we shouldn't need to copy it.

james-perretta commented on src/components/project_view/submission_detail/ag_suite_result.vue:
```diff
@@ -0,0 +1,283 @@
+  readonly CorrectnessLevel = CorrectnessLevel;
+
+  d_first_incorrect_setup: null | AGTestSuiteResultFeedback = null;
+  d_first_incorrect_case: null | AGTestCaseResultFeedback = null;
```
Is it possible to re-work these as computed properties?

james-perretta commented on src/components/project_view/submission_detail/mutation_suite_results.vue:
```diff
@@ -0,0 +1,169 @@
+    return CorrectnessLevel.not_available;
+  }
+
+  points_for_bugs_exposed_correctness(suite_result: MutationTestSuiteResultFeedback) {
```
Rename to `bugs_exposed_correctness`

james-perretta commented on src/components/project_view/submission_detail/ag_suite_result.vue:
```diff
@@ -0,0 +1,283 @@
+  d_loading = true;
+
+  d_ag_test_suite_result: AGTestSuiteResultFeedback | null = null;
+  d_fdbk_category: FeedbackCategory = FeedbackCategory.past_limit_submission;
```
I think we can get rid of this and just use fdbk_category.

james-perretta commented on src/components/project_view/submission_detail/ag_suite_result.vue:
```diff
@@ -0,0 +1,283 @@
+
+  d_loading = true;
+
+  d_ag_test_suite_result: AGTestSuiteResultFeedback | null = null;
```
I think we can get rid of this and just use ag_test_suite_result

james-perretta commented on src/components/project_view/submission_detail/ag_suite_result.vue:
```diff
@@ -0,0 +1,283 @@
+    this.d_ag_test_suite_result = this.ag_test_suite_result;
+    this.d_fdbk_category = this.fdbk_category;
+    this.determine_active_panels();
+    this.d_loading = false;
```
This function doesn't make any async calls, so we don't need to wrap the component in loading

james-perretta commented on src/components/project_view/submission_detail/ag_suite_result.vue:
```diff
@@ -0,0 +1,283 @@
+    return this.is_first_suite && this.d_first_incorrect_setup !== null;
+  }
+
+  case_result_correctness(case_result: AGTestCaseResultFeedback): CorrectnessLevel {
```
I think an approach like the one I suggested for mutation suite correctness (and what you do for return code and output correctness) will work here too:
- Points === 0 special case
- Put correctnesses in a array
- Filter out not_available, if nothing left then not_available
- Filter out info_only, if nothing left then info_only
- If everything left all_correct, all correct
- If everything left none_correct, none_correct
- Else, some_correct

james-perretta commented on src/components/project_view/submission_detail/ag_case_setup_result.vue:
```diff
@@ -0,0 +1,188 @@
+<template>
```
Can we name this file ag_suite_setup_result.vue instead? (Similarly for the component name)

james-perretta commented on src/components/project_view/submission_detail/ag_case_setup_result.vue:
```diff
@@ -0,0 +1,188 @@
+          <div class="feedback">
+            <div class="correctness-output">
+              <span v-if="setup_exit_status === ReturnCodeCorrectness.timed_out">
+                <i class="fas fa-clock timed-out-icon"></i>
```
Let's use the `far fa-clock` to be consistent with the other clocks

james-perretta commented on src/styles/components/submission_detail.scss:
```diff
@@ -0,0 +1,138 @@
+    border: 2px solid $white-gray;
+    padding: 10px 10px 4px 10px;
+    margin: 20px 0 20px 0;
+    border-radius: 5px;
```
Let's lower this border radius. Maybe 2px tops?

james-perretta commented on src/styles/components/submission_detail.scss:
```diff
@@ -0,0 +1,138 @@
+}
+
+.suite-result {
+    border: 2px solid $white-gray;
```
Can we use a darker border color?

james-perretta commented on src/components/project_view/submission_detail/submission_detail.vue:
```diff
@@ -0,0 +1,597 @@
+}
+
+#submission-detail {
+  padding: 9px 30px 30px 30px;
```
How much can we reduce the left and right padding?

james-perretta commented on src/components/project_view/submission_detail/ag_case_setup_result.vue:
```diff
@@ -0,0 +1,188 @@
+  d_setup_stderr: string | null = null;
+  d_setup_stdout_loaded = false;
+  d_setup_stderr_loaded = false;
+  d_ag_test_suite_result_output_size: ResultOutput.AGTestSuiteResultOutputSize | null = null;
```
Rename to `d_output_size`

james-perretta commented on src/components/project_view/submission_detail/ag_case_setup_result.vue:
```diff
@@ -0,0 +1,188 @@
+  d_setup_stdout_loaded = false;
+  d_setup_stderr_loaded = false;
+  d_ag_test_suite_result_output_size: ResultOutput.AGTestSuiteResultOutputSize | null = null;
+  d_ag_test_suite_result: AGTestSuiteResultFeedback | null = null;
```
Remove, use ag_test_suite_result directly

james-perretta commented on src/components/project_view/submission_detail/ag_case_setup_result.vue:
```diff
@@ -0,0 +1,188 @@
+  d_setup_stderr_loaded = false;
+  d_ag_test_suite_result_output_size: ResultOutput.AGTestSuiteResultOutputSize | null = null;
+  d_ag_test_suite_result: AGTestSuiteResultFeedback | null = null;
+  d_submission: Submission | null = null;
```
Remove, use submission_directly

james-perretta commented on src/components/project_view/submission_detail/ag_case_setup_result.vue:
```diff
@@ -0,0 +1,188 @@
+  }
+
+  @Watch('ag_test_suite_result')
+  async on_ag_test_suite_result_change(new_value: object, old_value: object) {
```
Use narrower type than `object`

james-perretta commented on src/components/project_view/submission_detail/ag_case_setup_result.vue:
```diff
@@ -0,0 +1,188 @@
+    await this.get_results();
+  }
+
+  async get_results() {
```
Rename to `get_output` or `load_output`

james-perretta commented on src/components/project_view/submission_detail/ag_case_setup_result.vue:
```diff
@@ -0,0 +1,188 @@
+    this.d_setup_stderr_loaded = true;
+  }
+
+  get setup_exit_status() {
```
Rename to `exit_status_correctness`?

james-perretta commented on src/components/project_view/submission_detail/ag_case_result.vue:
```diff
@@ -0,0 +1,164 @@
+    <template v-if="ag_test_case_result.ag_test_command_results.length > 1">
+      <div id="multi-command-body">
+        <div v-for="ag_test_command_result of ag_test_case_result.ag_test_command_results">
+          <submission-detail-panel
```
Add `:key="ag_test_command_result.pk"`

james-perretta commented on src/components/project_view/submission_detail/ag_case_result.vue:
```diff
@@ -0,0 +1,164 @@
+
+  readonly CorrectnessLevel = CorrectnessLevel;
+
+  command_result_correctness(command_result: AGTestCommandResultFeedback) {
```
Correctnesses array + filter approach here too?

If we use that approach, we might be able to come up with a "merge correctnesses" function to abstract away a decent chunk of that logic. (Don't try to do that just yet, let's get the individual correctness functions updated first and then see what that looks like)

james-perretta commented on src/components/project_view/submission_detail/ag_command_result.vue:
```diff
@@ -0,0 +1,387 @@
+<template>
```
Let's rename this to `ag_test_command_result.vue` because an AGCommand and an AGTestCommand aren't quite the same thing

james-perretta commented on src/components/project_view/submission_detail/ag_command_result.vue:
```diff
@@ -0,0 +1,387 @@
+            <div class="feedback">
+              <div class="short-output"
+                   id="expected-return-code">{{d_ag_test_command_result.expected_return_code
+                === ExpectedReturnCode.zero ? 0 : 'nonzero'}}</div>
```
Indentation here is kind of wonky. Maybe try:
```
                   id="expected-return-code">{{
                     d_ag_test_command_result.expected_return_code === ExpectedReturnCode.zero 
                     ? 0 : 'nonzero'}}</div>

```

or

```
              <div 
                class="short-output"
                id="expected-return-code"
              >{{d_ag_test_command_result.expected_return_code === ExpectedReturnCode.zero 
                 ? 0 : 'nonzero'}}</div>

```

james-perretta commented on src/components/project_view/submission_detail/ag_command_result.vue:
```diff
@@ -0,0 +1,387 @@
+  d_stderr_diff_loaded = false;
+  d_ag_test_command_result_output_size: ResultOutput.AGTestCommandResultOutputSize | null = null;
+  d_submission: Submission | null = null;
+  d_ag_test_command_result: AGTestCommandResultFeedback | null = null;
```
Remove, use ag_test_command_result

james-perretta commented on src/components/project_view/submission_detail/ag_command_result.vue:
```diff
@@ -0,0 +1,387 @@
+  d_stdout_diff_loaded = false;
+  d_stderr_diff_loaded = false;
+  d_ag_test_command_result_output_size: ResultOutput.AGTestCommandResultOutputSize | null = null;
+  d_submission: Submission | null = null;
```
Remove, use submission

james-perretta commented on src/components/project_view/submission_detail/ag_command_result.vue:
```diff
@@ -0,0 +1,387 @@
+  d_ag_test_command_result_output_size: ResultOutput.AGTestCommandResultOutputSize | null = null;
+  d_submission: Submission | null = null;
+  d_ag_test_command_result: AGTestCommandResultFeedback | null = null;
+  d_fdbk_category: FeedbackCategory = FeedbackCategory.past_limit_submission;
```
Remove, use fdbk_category

james-perretta commented on src/components/project_view/submission_detail/ag_command_result.vue:
```diff
@@ -0,0 +1,387 @@
+    await this.get_results();
+  }
+
+  async get_results() {
```
Rename to `get_output` or `load_output`? (use the name you chose in ag_case_setup_result.vue)

james-perretta commented on src/components/project_view/submission_detail/ag_case_result.vue:
```diff
@@ -0,0 +1,164 @@
+  }
+
+  command_result_output_correctness(command_result: AGTestCommandResultFeedback) {
+    let output_not_available = command_result.stdout_correct === null &&
```
It might be worth writing a helper function that gives you correctness for one set of output values and calling it twice in command_result_correctness

james-perretta commented on src/components/project_view/submission_detail/ag_case_result.vue:
```diff
@@ -0,0 +1,164 @@
+  command_result_correctness(command_result: AGTestCommandResultFeedback) {
+    let return_code_correctness = this.command_result_return_code_correctness(command_result);
+    let output_correctness = this.command_result_output_correctness(command_result);
+
```
I think this function should have a special case for total points <= 0? In that case it should be automatically none_correct. Note that If total points is negative, total points possible might be 0, so be sure to handle that case.

james-perretta commented on src/components/project_view/submission_detail/ag_case_setup_result.vue:
```diff
@@ -0,0 +1,188 @@
+          <div class="feedback-label"> Exit status: </div>
+          <div class="feedback">
+            <div class="correctness-output">
+              <span v-if="setup_exit_status === ReturnCodeCorrectness.timed_out">
```
Could we use checks for return code === null or timed out being true instead of using ReturnCodeCorrectness here?

james-perretta replied:
And then get rid of the setup_exit_status property?

james-perretta commented on src/components/project_view/submission_detail/mutation_suite_result.vue:
```diff
@@ -0,0 +1,798 @@
+              <span v-else-if="setup_return_code_correctness === ReturnCodeCorrectness.incorrect">
+                <i class="fas fa-times incorrect-icon"></i>
+              </span>
+              <span v-else-if="setup_return_code_correctness === ReturnCodeCorrectness.timed_out">
```
Similar to my comment in ag_case_setup_result.vue, can we use checks for return code === 0 and timed out being true instead of using ReturnCodeCorrectness as a middleman?

james-perretta commented on src/components/project_view/submission_detail/ag_command_result.vue:
```diff
@@ -0,0 +1,387 @@
+      <div id="exit-status-section">
+        <div class="feedback-row"
+             id="return-code-correctness-section"
+             v-if="return_code_correctness !== ReturnCodeCorrectness.not_available">
```
Replace with return_code_correct !== null

james-perretta commented on src/components/project_view/submission_detail/ag_command_result.vue:
```diff
@@ -0,0 +1,387 @@
+          </div>
+          <div class="feedback">
+            <div class="correctness-output">
+              <span v-if="return_code_correctness === ReturnCodeCorrectness.timed_out">
```
Replace with a truthiness check for timed_out

james-perretta commented on src/components/project_view/submission_detail/ag_command_result.vue:
```diff
@@ -0,0 +1,387 @@
+                <i class="far fa-clock timed-out-icon"></i>
+                <span class="timed-out-msg"> (Timed out) </span>
+              </span>
+              <span v-else-if="return_code_correctness === ReturnCodeCorrectness.correct">
```
Replace with truthiness check for return_code_correct

james-perretta commented on src/components/project_view/submission_detail/ag_command_result.vue:
```diff
@@ -0,0 +1,387 @@
+              <span v-else-if="return_code_correctness === ReturnCodeCorrectness.correct">
+                <i class="fas fa-check correct-icon"></i>
+              </span>
+              <span v-else-if="return_code_correctness === ReturnCodeCorrectness.incorrect">
```
Replace with `return_code_correct !== null && !return_code_correct`

james-perretta commented on src/components/project_view/submission_detail/ag_command_result.vue:
```diff
@@ -0,0 +1,387 @@
+                            {'actual-return-code-correct': return_code_correctness
+                             === ReturnCodeCorrectness.correct},
+                            {'actual-return-code-incorrect': return_code_correctness
+                             === ReturnCodeCorrectness.incorrect}]"
```
Combine into one dictionary.
Use the same expressions as above instead of the return_code_correctness getters. You can make a getter for `return_code_incorrect` so that you don't have to duplicate the null check.

james-perretta commented on src/components/project_view/submission_detail/ag_command_result.vue:
```diff
@@ -0,0 +1,387 @@
+           === ValueFeedbackLevel.expected_and_actual;
+  }
+
+  get return_code_correctness() {
```
After removing references to this property, remove it.

james-perretta commented on src/components/project_view/submission_detail/return_code_correctness.ts:
```diff
@@ -0,0 +1,6 @@
+export enum ReturnCodeCorrectness {
```
Hoping that I addressed all uses of this enum in other comments, remove this enum and this file.

ashberg committed:
Made the changes requested on the pull request for branch 105 pertaining to the ag_suite_setup_result component.

ashberg committed:
Making the requested changes to the MutationSuiteResults component.

ashberg committed:
Making the requested changes to the AGTestCommandResult component.

ashberg committed:
Removed the ReturnCodeCorrectness enum and file of the same name. Removed references to ReturnCodeCorrectness from the mutation_suite_result component.

ashberg committed:
Reworked d_first_incorrect_setup and d_first_incorrect case as computed properties.

ashberg committed:
Made some style-related changes to the submission_detail component.

ashberg committed:
Making the changes requested in the pull request for branch 105 pertaining to the mutation_suite_result component.

ashberg committed:
Updated the case_result_correctness function in ag_suite_result to use the correctness array filtering method like in mutation_suite_results.

ashberg committed:
Adjusted tests in test_mutation_suite_results.ts and test_ag_case_result.ts.

ashberg committed:
Adjusted the command_result_output_correctness function so that it could be called twice in command_result_correctness with stderr_correct and stdout_correct as arguments.

ashberg committed:
Added the ViewFile component to AGCaseSetupResult, AGTestCommandResult and MutationSuiteResult.

ashberg committed:
Added an extra condition for loading the diff contents for stdout_diff and stderr_diff in ag_test_command_result.

james-perretta committed:
Changed AGSuiteResult.first_incorrect_setup to setup_is_incorrect.
Value is now a bool instead of nullable, since an AG suite result can only ever have one setup command.
Also fixed some name inconsistencies in test names in test_ag_suite_result.ts.

james-perretta committed:
Minor logic consolidation in mutation_suite_result.vue

james-perretta committed:
Changed some try-finallys to use the toggle function in submission_detail.vue.

james-perretta committed:
Changed submission-detail-panel to result-panel

james-perretta committed:
Tweaked waiting_for_deferred/finished_grading message.
Made that message show up when points are not shown.

james-perretta committed:
Fixed css class name in submission list.

james-perretta committed:
Fixed build errors

ashberg committed:
Added a check_all_command_correctnesses function to test_ag_suite_result.ts.

ashberg committed:
Renamed 'check_all_command_correctnesses' to 'check_all_case_correctness_levels'.

ashberg committed:
Added a make_wrapper function to test_ag_suite_result.ts to call to mount the component. Removed the check_command_result_correctness function from test_ag_suite_result.ts

james-perretta committed:
Wrote make_wrapper function to reduce duplication in test_mutation_suite_result.ts

ashberg committed:
Added a make_wrapper function to test_submission_detail.ts.
